### DISCUSSION TITLE
How to boot bare metal on actual hardware?

### DISCUSSION THREAD

**Discussion by RaymiiOrg** [jart/cosmopolitan#805] [Q&A]
I've compiled a Vi clone (https://raymii.org/s/blog/Bare_Metal_Boot_to_Vi.html) with cosmopolitan, which works on Linux and QEMU. I want to run that on actual hardware, but I'm not sure how to proceed. I've tried `dd`-ing the image to an USB drive, which did not show up in my BIOS boot menu. I also wrote the image to an actual floppy drive using `dd` and tried to boot, again did not show up in my BIOS boot menu. I don't have access to an UEFI enabled machine. Are there instructions available to run a compiled executable on actual hardware, not QEMU?

**Comment by tkchia** [comment]
Hello @RaymiiOrg,

`dd`-ing _should_ work.  Are you able to get one of the example programs in `example/` to run on an actual machine?  Try e.g.

> `sudo dd if=o//examples/vga.com of=/dev/sdX conv=notrunc`

(replace `/dev/sdX` with your device's name) and then reboot with the USB drive.

If `vga.com` runs but `vi.com` does not run, then perhaps the issue was in how `vi.com` was built.  If `vga.com` does not run on your machine either, then that is very strange.  But maybe try booting `vga.com` first.

(Anyway, Cosmopolitan's bare metal support defaults to only supporting legacy BIOS machines actually.  You need extra steps in your program to support UEFI machines &mdash; take a look at `examples/vga2.c`.)

Thank you!
